pilot/pkg/networking/core/v1alpha3/gateway_test.go

```diff
@@ -946,6 +947,75 @@ func TestGatewayHTTPRouteConfig(t *testing.T) {
 
 }
 
+func TestBuildGatewayListeners(t *testing.T) {
+	cases := []struct {
+		name              string
+		node              *pilot_model.Proxy
+		gateway           *networking.Gateway
+		expectedListeners []string
+	}{
+		{
+			"targetPort overrides service port",
+			&pilot_model.Proxy{
+				ServiceInstances: []*pilot_model.ServiceInstance{
+					{
+						Service: &pilot_model.Service{
+							Hostname: "test",
+						},
+						ServicePort: &pilot_model.Port{
+							Port: 80,
+						},
+						Endpoint: &pilot_model.IstioEndpoint{
+							EndpointPort: 8080,
+						},
+					},
+				},
+			},
+			&networking.Gateway{
+				Servers: []*networking.Server{
+					{
+						Port: &networking.Port{Name: "http", Number: 80, Protocol: "HTTP"},
+					},
+				},
+			},
+			[]string{"0.0.0.0_8080"},
```

Comment:
For ingress gateway, there is no inbound listener, so the target port does not take effect for it.

Reply:
Hi @hzxuzhonghu, is this a problem with the test I wrote or the implementation? I tested this and it works, it creates a listener on the `targetPort` instead of the port

Reply:
The listener of gateway should have nothing to do with the targetport.

Reply:
I don't agree, that's the whole point of the change: to have an `ingress-gateway` service with `port: 80` and `targetPort: 8080`, so that an `Ingress` resource can still point to port 80 of the ingress-gateway service, but the pod running the gateway listens on port 8080, which does not require privileges.

Reply:
We currently assume in the code that `port` and `targetPort` for services targeting gateways are always the same, which they don't have to be.

Reply:
Understood now